### PR TITLE
zap-full-scan.py - first version

### PR DESCRIPTION
**Pull request by psiinon**

### CODE REVIEW

File: build/docker/zap-full-scan.py [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+#	3:	Any other failure
+# By default all alerts found by ZAP will be treated as WARNings.
+# You can use the -c or -u parameters to specify a configuration file to override
+# this.
+# You can generate a template configuration file using the -g parameter. You will
+# then need to change 'WARN' to 'FAIL', 'INFO' or 'IGNORE' for the rules you want
+# to be handled differently.
+# You can also add your own messages for the rules by appending them after a tab
+# at the end of each line.
+# The active scan rules run are currently hardcoded in the API-Minimal.policy file
```
Comment (thc202) [inline build/docker/zap-full-scan.py]: Should say that it uses the Default Policy?

File: build/docker/zap-full-scan.py [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+# Scan rules that aren't really relevant, eg the examples rules in the alpha set
+blacklist = ['-1', '50003', '60000', '60001']
+
+# Scan rules that are being addressed
+in_progress_issues = {}
+
+logging.basicConfig(level=logging.INFO, format='%(asctime)s %(message)s')
+# Hide "Starting new HTTP connection" messages
+logging.getLogger("requests").setLevel(logging.WARNING)
+
```
Comment (thc202) [inline build/docker/zap-full-scan.py]: Should have another empty line.

File: build/docker/zap-full-scan.py [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+def usage():
+    print('Usage: zap-full-scan.py -t <target> [options]')
```
Comment (thc202) [inline build/docker/zap-full-scan.py]: Should mention the `-j` arg.

File: build/docker/zap-full-scan.py [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+    generate = ''
+    port = 0
+    detailed_output = True
+    report_html = ''
+    report_md = ''
+    report_xml = ''
+    target = ''
+    target_file = ''
+    target_url = ''
+    format = ''
```
Comment (thc202) [inline build/docker/zap-full-scan.py]: This (and above two) could be removed, does not seem to be used.

File: build/docker/zap-full-scan.py [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+# at the end of each line.
+# The active scan rules run are currently hardcoded in the API-Minimal.policy file
+# the plan is to make these configurable in a future version of this script.
+
+import getopt
+import json
+import logging
+import os
+import os.path
+import subprocess
```
Comment (thc202) [inline build/docker/zap-full-scan.py]: Could be removed, does not seem to be used.

File: build/docker/zap-full-scan.py [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+
+        except OSError:
+            logging.warning('Failed to start ZAP :(')
+            sys.exit(3)
+
+    else:
+        # Not running in docker, so start one
+        mount_dir = ''
+        if context_file:
+            mount_dir =  os.path.dirname(os.path.abspath(context_file))
```
Comment (thc202) [inline build/docker/zap-full-scan.py]: Two spaces after equals.

File: build/docker/zap-full-scan.py [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+
+        if (ajax):
+            zap_ajax_spider(zap, target, mins)
+
+        if (delay):
+            start_scan = datetime.now()
+            while ((datetime.now() - start_scan).seconds < delay):
+                time.sleep(5)
+                logging.debug('Delay active scan ' + str(delay -(datetime.now() - start_scan).seconds) + ' seconds')
+
```
Comment (thc202) [inline build/docker/zap-full-scan.py]: Whitespace chars could be removed.

File: build/docker/zap_common.py [hunk 8]
```diff
@@ -254,6 +254,18 @@ def zap_ajax_spider(zap, target, max_time):
     logging.debug('Ajax Spider complete')
 
 
+def zap_active_scan(zap, target, policy):
```
Comment (thc202) [inline build/docker/zap_common.py]: API script could be changed to use this function.

**Comment by psiinon** [comment]
Hopefully fixed those issues

**Comment by thc202** [comment]
Thank you!

File: build/docker/zap_common.py [hunk 9]
```diff
@@ -254,6 +254,18 @@ def zap_ajax_spider(zap, target, max_time):
...
+    logging.debug('Active Scan ' + target + ' with policy ' + policy)
+    ascan_scan_id = zap.ascan.scan(target, recurse=True, scanpolicyname=policy)
+    time.sleep(5)
+
+    while(int(zap.ascan.status(ascan_scan_id)) < 100):
+        logging.debug('Active Scan progress %: ' + zap.ascan.status(ascan_scan_id))
+        time.sleep(5)
+    logging.debug('Active Scan complete')
+    logging.debug(zap.ascan.scan_progress(ascan_scan_id))
```
Comment (kingthorin) [inline build/docker/zap_common.py]: TBH I haven't worked on any Python in ages and don't remember the syntax. But something has to be off here. Either the sleep isn't within the while or the loggin.debug is and should be indented more?

Comment (psiinon) [inline build/docker/zap_common.py]: @kingthorin both lines (263 and 264) are indented the same amount and so are both in the while loop. So I _think_ its as its supposed to be...

**Pull request merged**